docs/source/dev.rst

```diff
@@ -4,13 +4,142 @@
 
    The full license is in the file LICENSE, distributed with this software.
 
-Developer section
-=================
+Internal of xeus
+================
 
-For developers who want to contribute to `xeus` or implement their own kernel
-and not use the default one, here is how the default kernel is implemented:
+`xeus` is internally architected around three main components:
 
-The kernel contains the server and the interpreter,
+.. image:: xeus_archi.svg
+   :align: center
+
+- The server is the middleware component responsible for receiving and transmitting messages to
+  the Jupyter client. It is build upon `ZeroMQ` and handles the threads of the applicaiton.
+- The kernel core routes the messages to the right method of the interpreter and performs
+  some common operations such as storing the message and its answer in the history
+  manager, or sending messages to publish to the server.
+- The interpreter provides the interface that kernel authors must implement.
+
+The interpreter and the server are well isolated from each other, only the kernel core can
+interact with them. The kernel core is also loosely coupled to the server, which makes it
+easy to replace the server implementation provided by `xeus` with a custom one.
+
+Server
+------
+
+Public API
+~~~~~~~~~~
+
+The server part of `xeus` provides a public API made of:
+
+- ``xserver.hpp``: This file contains the base class ``xserver``, which must be inherited from any
+  class implementing a server. This is the unique entry point into the server component used by the
+  kernel core.
+- ``xerver_zmq.hpp``: This file contains the interface of the default server implementation, that can
+  be used directly or extended in order to override parts of its behavior.
+
+Before we dive into the details of the server implementation, let's have a look at the public interface:
+
+.. code::
```

Comment:
Maybe you can use 

```
.. literalinclude:: ../../include/xeus/xserver.hpp
   :language: cpp
   :dedent: 4
   :lines: 22-75
```

instead of copying the code in the documentation. So that we don't have to update it.

I don't really like the fact that it depends on the line numbers but that's the best I could find.